### DISCUSSION TITLE
Duplicates on image SBOM (SPDX) report

### DISCUSSION THREAD

**Discussion by juan131** [aquasecurity/trivy#6204] [Bugs]
### Description

Scanning to report SBOM (`trivy image --format spdx-json ...`) include duplicated app libraries when they're detected during analysis twice (e.g. it was both detected with "jar" analyzer and "sbom" analyzer).

### Desired Behavior

Libraries shouldn't be included twice in the SPDX output.

### Actual Behavior

Libraries are duplicated when scanning certain images.

### Reproduction Steps

1. Run `trivy image --format spdx-json --output sbom.json bitnami/postgresql:15.5.0-debian-11-r21`
2. Inspect `sbom.json` (e.g. execute: `cat sbom.json | jq -r '.packages[] | select(.name == "org.postgresql:pljava")'`) to find duplicates

The library is detected twice due to the "jar" analyzer + the "bitnami" analyzer based on the `.spdx-postgresql.spdx` included in the container image.

### Target

Container Image

### Scanner

None

### Output Format

SPDX

### Mode

Standalone

### Debug Output

```bash
It adds no value
```

### Operating System

macOS sonoma

### Version

```console
$ trivy --version
Version: 0.49.1
Vulnerability DB:
  Version: 2
  UpdatedAt: 2024-01-09 12:13:28.274214364 +0000 UTC
  NextUpdate: 2024-01-09 18:13:28.274214053 +0000 UTC
  DownloadedAt: 2024-01-09 15:56:37.96018 +0000 UTC
Java DB:
  Version: 1
  UpdatedAt: 2024-02-26 05:38:21.591482531 +0000 UTC
  NextUpdate: 2024-02-29 05:38:21.591482331 +0000 UTC
  DownloadedAt: 2024-02-26 11:17:18.334661 +0000 UTC
```

### Checklist

- [X] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by juan131** [comment]
Similar although not the same as https://github.com/aquasecurity/trivy/issues/5812

**Comment by juan131** [comment]
Duplicates can be removed by applying the diff below at `pkg/fanal/applier/docker.go` but it's very likely not the best solution:

```diff
$ git diff
diff --git a/pkg/fanal/applier/docker.go b/pkg/fanal/applier/docker.go
index 730737e8..8b07aef5 100644
--- a/pkg/fanal/applier/docker.go
+++ b/pkg/fanal/applier/docker.go
@@ -277,7 +277,19 @@ func aggregate(detail *ftypes.ArtifactDetail) {
                        apps = append(apps, app)
                        continue
                }
-               a.Libraries = append(a.Libraries, app.Libraries...)
+               for _, newLib := range app.Libraries {
+                       found := false
+                       log.Logger.Infof("NewLib: %s", newLib)
+                       for _, existingLib := range a.Libraries {
+                               if lo.FromPtr(existingLib.Identifier.PURL).String() == lo.FromPtr(newLib.Identifier.PURL).String() {
+                                       found = true
+                                       break
+                               }
+                       }
+                       if !found {
+                               a.Libraries = append(a.Libraries, newLib)
+                       }
+               }
        }

        for _, app := range aggregatedApps {
```

cc @DmitriyLewen

**Reply by DmitriyLewen** [reply]
Hello @juan131 
I also thought like you, but after recheck i understood that these are not duplicate packages (different filepaths):
```
     {
      "name": "org.postgresql:pljava",
      "SPDXID": "SPDXRef-Package-87c42734dcc2db36",
     ...
     },
    {
      "spdxElementId": "SPDXRef-Package-87c42734dcc2db36",
      "relatedSpdxElement": "SPDXRef-File-730c3112cf56f202",
      "relationshipType": "CONTAINS"
    },
    {
      "fileName": "opt/bitnami/postgresql/pljava-1.6.6.jar",
      "SPDXID": "SPDXRef-File-730c3112cf56f202",
      "checksums": null,
      "copyrightText": ""
    },
```
```
    {
      "name": "org.postgresql:pljava",
      "SPDXID": "SPDXRef-Package-e2967b0035e3b94a",
      ...
    },
    {
      "spdxElementId": "SPDXRef-Package-e2967b0035e3b94a",
      "relatedSpdxElement": "SPDXRef-File-aa3933db31ba49d3",
      "relationshipType": "CONTAINS"
    },
    {
      "fileName": "opt/bitnami/postgresql/share/pljava/pljava-1.6.6.jar",
      "SPDXID": "SPDXRef-File-aa3933db31ba49d3",
      "checksums": [
        {
          "algorithm": "SHA1",
          "checksumValue": "227031a762316e6a006d9c7f4f473b05d93c6e90"
        }
      ],
      "copyrightText": ""
    },
```

**Reply by juan131** [reply]
Awesome, thanks for looking into this.

**Comment by juan131** [comment]
Another weird thing we detected is that Trivy is not properly generating the checksum for the file associated to package, see:

```console
$ cat sbom.json | jq -r '.files[] | select(.fileName == "opt/bitnami/postgresql/pljava-1.6.6.jar")'
{
  "fileName": "opt/bitnami/postgresql/pljava-1.6.6.jar",
  "SPDXID": "SPDXRef-File-730c3112cf56f202",
  "checksums": null,
  "copyrightText": ""
}
```

**Reply by DmitriyLewen** [reply]
hm... looks like we don't move checksums from SPDX file to `Application`.
Take a look this test:
https://github.com/aquasecurity/trivy/blob/64926d8423d1e69ab4bde51f1e59ed56960f9bcb/pkg/sbom/spdx/unmarshal_test.go#L165-L191

**Reply by DmitriyLewen** [reply]
I will try to find a solution to this case when I have time and write to you.

**Reply by DmitriyLewen** [reply]
Hello @juan131 
We are working on updating SBOM logic - #6240.
The new logic [saves](https://github.com/aquasecurity/trivy/blob/7da791118775e8185098c438934851271f75f80e/pkg/sbom/io/decode.go#L218-L220) checksums from SBOM.

The new PR will add `BOM` support for `SPDX`. At the same time, the problem with checksums should be fixed.

**Reply by juan131** [reply]
Awesome, thanks so much @DmitriyLewen
